de_web_plugin.cpp

```diff
@@ -9781,6 +9781,23 @@ Sensor *DeRestPluginPrivate::getSensorNodeForAddressAndEndpoint(const deCONZ::Ad
 
 }
 
+/*! Returns the first Sensor for its given \p Address and \p Endpoint and \p Cluster or 0 if not found.
+ */
+Sensor *DeRestPluginPrivate::getSensorNodeForAddressEndpointAndCluster(const deCONZ::Address &addr, quint8 ep, quint16 cluster)
+{
+    for (Sensor &sensor: sensors)
+    {
+        if (sensor.deletedState() != Sensor::StateNormal || !sensor.node())                                     { continue; }
+        if (!((sensor.address().hasExt() && addr.hasExt() && sensor.address().ext() == addr.ext()) ||
+            (sensor.address().hasNwk() && addr.hasNwk() && sensor.address().nwk() == addr.nwk())))              { continue; }
+        if (!((sensor.fingerPrint().hasInCluster(cluster) || sensor.fingerPrint().hasOutCluster(cluster))))     { continue; }
+        if (sensor.fingerPrint().endpoint != ep)                                                                { continue; }
+        else { return &sensor; }
```

Comment:
I'm a big fan of guard expressions but the more complex ones are a bit hard to reason about :)

*Live code review* 😄 

Here is what I did to understand what the code does since the address check gives me a hard time:

```cpp
if (sensor.deletedState() != Sensor::StateNormal || !sensor.node())                                     { continue; }
if (!
      (
         (sensor.address().hasExt() && addr.hasExt() && sensor.address().ext() == addr.ext()) ||
         // FIXME When both ext() are present but different the nwk() check could still success (albeit unlikely)
         (sensor.address().hasNwk() && addr.hasNwk() && sensor.address().nwk() == addr.nwk())
      )
   )
{ continue; }

if (!
       (
          (sensor.fingerPrint().hasInCluster(cluster) || sensor.fingerPrint().hasOutCluster(cluster))
       )
   )
{ continue; }

if (sensor.fingerPrint().endpoint != ep)
{ continue; }

else { return &sensor; }
```

The difficulty for me when reading code in general is to keep lots of logic and parantheses in my head to follow what is actually happening 😄 

Another attempt, since the address check can be reused in other parts of the code, we can made it a extra function `isSameAddress(a, b)` which is a bit stricter and testable. In `getSensorNodeForAddressEndpointAndCluster()` it becomes one thing less to think about.

```cpp
/*! Compare addresses where either NWK or MAC address might not be known.
    \returns true if both adresses have same MAC address (strong guaranty).
    \returns true if at least one of the addresses doesn't have MAC but the NWK addresses are equal.
    \returns false otherwise.
*/
bool isSameAddress(const deCONZ::Address &a, const deCONZ::Address &b)
{
    if (a.hasExt() && b.hasExt())
    {
        // nested if statement, so the NWK check won't be made if both MAC addresses are known
        if (a.ext() != b.ext()) 
        {
             return false;
        }
    }
    else  if (a.hasNwk() && b.hasNwk())
    {
       if (a.nwk() != b.nwk())
       {
            return false;
       }
    }
    else {  return false; }

    return true;
}

/*! Returns the first Sensor for its given \p Address and \p Endpoint and \p Cluster or nullptr if not found.
 */
Sensor *DeRestPluginPrivate::getSensorNodeForAddressEndpointAndCluster(const deCONZ::Address &addr, quint8 ep, quint16 cluster)
{
    for (Sensor &sensor: sensors)
    {
        if (sensor.deletedState() != Sensor::StateNormal || !sensor.node())  { continue; }
        if (!isSameAddress(sensor.address(), addr))    {  continue; }
        if (sensor.fingerPrint().hasInCluster(cluster) || sensor.fingerPrint().hasOutCluster(cluster)) { }
        else { continue; }

        if (sensor.fingerPrint().endpoint != ep) { continue; }
        else { return &sensor; }
    }
    return nullptr;
}
```

Reply:
@manup at times, you really give me a hard time to find the right balance in what to keep together and what to split :stuck_out_tongue_winking_eye: Readability and to not have to twist your thoughts was the initial reason why the code we discussed via PM was longer, let aside the positive/negative checks.

Regardless, I'm all in here but it also raises the question, why we don't take this one step further and consider dissecting all the related `getSensorNodeForXYZ()` we have to create something universal. Maybe by using an additional argument to jump out or so. The other functions are quite similar in code usage, but yet different. What do you think?

What still disturbs me is that there's still a chance you'd need to have more than one sensor (e.g. for Multisensors having one endpoint and updating battery).

Reply:
Ya I know :) lately I'm becoming more aware that splitting up functions helps a lot to reason about them — at least if they really just do what the function name implies. Digging deeper into unit testing it's neat to be able to test small functions as well as larger ones which use these.

>Regardless, I'm all in here but it also raises the question, why we don't take this one step further and consider dissecting all the related getSensorNodeForXYZ() we have to create something universal. Maybe by using an additional argument to jump out or so. The other functions are quite similar in code usage, but yet different. What do you think?

I'm all in, personally I'd like them to be free standing functions which take the `sensors`container as parameter.
For example the above could be:

```cpp
Sensor *getSensorNodeForAddressEndpointAndCluster(std::vector<Sensor> &sensors,
                                                  const deCONZ::Address &addr,                                                  
                                                  quint8 ep, quint16 cluster)
```

Mainly to decouple this from the plugin and being able to write some tests.

I'm not sure if it is possible for the current code base to create a universial `getSensor(whatGoesInHere?)` function since depending on the context the needed and available "matching arguments" are different. But it would be great if we can figure the contexts out and reduce the getSensorXYZ() functions to the most precise ones.

>What still disturbs me is that there's still a chance you'd need to have more than one sensor (e.g. for Multisensors having one endpoint and updating battery).

Yes tricky with the current code base without lots of checks, the Device code doesn't have this problem anymore as commands are just directly forwarded to where they belong without extra code.